python/phase-1.md

```diff
@@ -0,0 +1,441 @@
+    likely, we'll need a typing Protocol that covers both.
+    * Due to how Core wraps gRPC today, we are probably going to have to have every call in
+      [WorkflowService](https://github.com/temporalio/api/blob/master/temporal/api/workflowservice/v1/service.proto)
+      manually wrapped (though the protos will still be dynamic). This should not pose a huge problem since the RPC
+      count on that service rarely changes.
+* We are intentionally defining a `Client` instead of a `WorkflowClient` since this will be general purpose
+  * Having a general purpose client keeps things simple, but this will probably only ever contain the pieces above (and
+    overloads thereof) and async activity heartbeat/complete
+* We call it `start_workflow` instead of just `start` to be clearer
+* We have an `execute_workflow` to combine start + result waiting
+* We have embedded `start_signal` and `start_signal_args` as parameters, so if the former is present we will perform a
+  signal with start. Is this ok?
+  * After discussion, it was determined this is acceptable
+* After discussion, we determined that the concept of a "default task queue" at the client level is not worth it right
+  now
+
+### Worker
+
+Exported from `temporalio` (but probably in `temporalio/worker/worker.py`, exported from `temporalio.worker` then
+exported from `temporalio`):
+
+```python
+class Worker:
+    def __init__(
+        self,
+        core: temporalio.Core,
+        *,
+        task_queue: str,
+        # Must be a mapping with the key as the activity name and the value as a
+        # callable to be called when the activity is invoked.
+        activities: Mapping[str, Callable[P, T]]={},
+        # An executor to be used for synchronous activities. If set, the worker
+        # does not shut this down. If unset, the worker may use a thread pool
+        # internally.
+        activity_executor: Optional[concurrent.futures.Executor]=None,
+        # Whether to use the given/defaulted executor for async activities as
+        # well.
+        use_activity_executor_on_async: bool = False,
+        # ...
+    ) -> None:
+        pass
+
+    async def run() -> None:
+        pass
+```
+
+Notes:
+
+* We have left off workflow registration until a later phase
+* Stopping the worker is done via cancellation of the run task/future
+* The constructor will fail if the task queue has already been registered with that instance of core
+* Internally, activities will run in the same event loop `run()` is called on
+* We allow a custom executor and potentially even default one for sync activities
+  * Should we default an executor for sync?
+  * Is it ok to not default an executor for async? Or should we use it for them too?
+* Should we have an `__aenter__` and an `__aexit__`?
+  * Probably, and change the examples to use `with` to ensure worker is shut down properly
+  * This means we have to treat those two calls as separate start and shutdown
```

Comment:
Doesn't look like this would work here.
What would be in the body of the `with` statement? It'd have to block on something.

Reply:
Added `start` and `shutdown`